src/prefect/utilities/processutils.py

```diff
@@ -311,28 +312,42 @@ async def stream_text(source: TextReceiveStream, sink: Optional[TextSink]):
             raise TypeError(f"Unsupported sink type {type(sink).__name__}")
 
 
-def kill_on_interrupt(pid: int, process_name: str, print_fn: Callable):
-    """Kill a process with the given `pid` when a SIGNINT is received."""
+def forward_signal_handler(
+    pid: int, signum: int, *signums: int, process_name: str, print_fn: Callable
+):
+    """Forward subsequent signum events (e.g. interrupts) to respective signums."""
+    current_signal, future_signals = signums[0], signums[1:]
+
+    def handler(*args):
+        print_fn(
+            f"Received {getattr(signum, 'name', signum)}. "
+            f"Sending {getattr(current_signal, 'name', current_signal)} to"
+            f" {process_name} (PID {pid})..."
+        )
+        os.kill(pid, current_signal)
+        if future_signals:
+            forward_signal_handler(
+                pid,
+                signum,
+                *future_signals,
+                process_name=process_name,
+                print_fn=print_fn,
+            )
 
-    # In a non-windows enviornment first interrupt with send a SIGTERM, then
-    # subsequent interrupts will send SIGKILL. In Windows we use
-    # CTRL_BREAK_EVENT as SIGTERM is useless:
-    # https://bugs.python.org/issue26350
-    if sys.platform == "win32":
+    # register current and future signal handlers
+    signal.signal(signum, handler)
 
-        def stop_process(*args):
-            print_fn(f"\nStopping {process_name}...")
-            os.kill(pid, signal.CTRL_BREAK_EVENT)
 
+def setup_signal_handlers_server(pid: int, process_name: str, print_fn: Callable):
+    """Handle interrupts of the server gracefully."""
+    setup_handler = partial(
+        forward_signal_handler, pid, process_name=process_name, print_fn=print_fn
+    )
+    # https://bugs.python.org/issue26350
+    if sys.platform == "win32":
+        setup_handler(signal.SIGINT, signal.CTRL_BREAK_EVENT)
```

Comment:
Should we retain the comment from before?

Reply:
done!